File: Python/happyprime/HappyPrime.java
import java.util.Arrays;
import java.util.HashSet;
import java.util.Scanner;

public class HappyPrime {
   private static boolean[] prime = new boolean[10_001];
   public static void main(String[] args) {
      prime();
      Scanner in = new Scanner(System.in);

      int cases = in.nextInt();

      for (int i = 0; i < cases; i++) {
         in.nextInt();
         Integer num = in.nextInt();

         if (prime[num] == false) {
            System.out.println((i + 1) + " " + num + " NO");
            continue;
         }
         else {
            if (happy(num, new HashSet<Integer>())) {
               System.out.println((i + 1) + " " + num + " YES");
            } else {
               System.out.println((i + 1) + " " + num + " NO");
            }
         }
      }
   }

   private static void prime() {
      Arrays.fill(prime, true);
      for (int i = 0; i < 10_000; i++) {
         if (prime[i] == false)
            continue;
         if (i <= 1) {
            prime[i] = false;
            continue;
         }

         for (int mults = i*i; mults < 10_000; mults += i) {
            prime[mults] = false;
         }
      }

      prime[10_000] = false;
      prime[0] = false;
      prime[1] = false;
   }

   private static boolean happy(Integer num, HashSet<Integer> seen) {
      int total = 0;

      String[] nums = num.toString().split("");

      for (String curr : nums) {
         total += Integer.parseInt(curr) * Integer.parseInt(curr);
      }

      if (total == 1)
         return true;
      else if (seen.contains(total))
         return false;
      else {
         seen.add(total);
         return happy(total, seen);
      }
   }
}

File: Python/happyprime/HappyPrime.py
def happy(num, seen):
   total = 0
      
   for curr in list(str(num)):
      total += int(curr)**2
      
   if total == 1:
      return True
   elif total in set(seen):
      return False
   else:
      set(seen).add(total)
      return happy(str(total), set(seen))

prime = [True for idk in range(10001)]

for x in range(10000):
   if prime[x] == False:
      continue
   if x <= 1:
      prime[x] = False
      continue
   
   for mults in range(x**2, 10000, x):
      prime[mults] = False
      
prime[10000] = False
prime[0] = False
prime[1] = False
      
cases = int(input())

for idk in range(cases):
   line = input().split()
   num = int(line[1])
   
   if prime[num] == False:
      print((idk + 1), num, "NO")
      continue
   else:
      if bool(happy(str(num), set())):
         print((idk + 1), num, "YES")
      else:
         print((idk + 1), num, "NO")